### DISCUSSION TITLE
Accessing POST data in getServerSideProps

### DISCUSSION THREAD

**Discussion by bradders** [vercel/next.js#12152] [Help]
What's the quickest/easiest way to access POST data from inside `getServerSideProps()`?

**Comment by leerob** [comment]
What are you trying to accomplish? You can probably just use `fetch`. 

```js
const response = await fetch(
  'URL',
  {
    body: JSON.stringify({ body }),
    headers: {
      Authorization: `Token ${API_KEY}`,
      'Content-Type': 'application/json'
    },
    method: 'POST'
  }
);
```

**Reply by bradders** [reply]
I'm POSTing to my route, so I was wondering how to access my POST'ed data from within getServerSideProps. Your example above makes sense, but my concern is populating that `body` variable you've got.

I can't see anything in the context res or req (passed into getServerSideProps) that would help me out. Unless I'm missing something?

**Reply by leerob** [reply]
When you say "my route", are you talking about an API route? If so, then it [will not work](https://nextjs.org/docs/basic-features/data-fetching#write-server-side-code-directly).

> You should not fetch an API route from `getStaticProps` — instead, you can write the server-side code directly in `getStaticProps`.

**Reply by bradders** [reply]
That's getStaticProps though, right? I'm trying to do things SSR on the server only in getServerSideProps

**Reply by leerob** [reply]
Whoops, yep. Sorry about that. Should be able to do this.

```js
export async function getServerSideProps() {
  // Fetch data from external API
  const res = await fetch(`https://.../data`)
  const data = await res.json()

  // Pass data to the page via props
  return { props: { data } }
}
```

**Reply by ks-himanshub** [reply]
Hey lee,

I want to communicate with `getServerSideProps` , right now I can do it with `query` params that we get in props of it.

but that is adding query in my browser URL, like submitting a form with GET method

can we do it without query in my browser, like using POST method for it

righ now  I am using Pages Router version and `useRouter` from next/router

**Comment by rafde** [comment]
```tsx
import qs from 'querystring';
import { GetServerSideProps, } from 'next';

export async function getServerSideProps( context: GetServerSideProps ) {
	const props = {
		username: '',
		password: '',
	};
	const {
		req,
	} = context;

	const streamPromise = new Promise( ( resolve, reject ) => {
		let postBody = '';

		req.on( 'data', ( data ) => {
			// convert Buffer to string
			postBody += data.toString();
		} );

		req.on( 'end', () => {
			const postData = qs.parse( postBody );
			resolve(postData);
		} );
	} );

	try {
		const {
			username,
			password,
		} = await streamPromise;
		props.username = username;
		props.password = password;
	} catch ( error ) {
		
	}

	return {
		// will be passed to the page component as props
		props,
	};
}
```
not sure what you want to do with the data, but there you go

**Reply by choejoonkyung** [reply]
thx bro.

**Comment by ericclemmons** [comment]
I found that if I `POST` a page, `req.method` correctly has `POST`, but `req.body` is `undefined`.

The gist is that traditional HTTP form posts should work with `getServerSideProps`.

**Comment by lvpaspaltzis** [comment]
`req` is `http.IncomingMessage` so I have managed to read the body data by using a listener:

```typescript
export const getServerSideProps: GetServerSideProps = async ({req}) => {
  if (req.method == "POST") {
    let body = ''
    req.on('data', (chunk) => {
      body += chunk
    })
    req.on('end', () => {
      console.log(body);
  });
  }
  return { props: {} };
};
```

**Reply by lvpaspaltzis** [reply]
even easier using `raw-body`:

```typescript
import getRawBody from 'raw-body';

...

export const getServerSideProps: GetServerSideProps = async ({req}) => {
  if (req.method == "POST") {
    const body = await getRawBody(req)
    console.log(body.toString("utf-8"))
  }
  
  return { props: {} };
};
```

**Reply by xstable** [reply]
Awesome Solution, thanks a lot!

**Reply by manuschillerdev** [reply]
thanks for the solution!
It feels like `req.body` should be filled by default in `getServerSideProps` without any workarounds. Is there a reason why it is not?

**Reply by lvpaspaltzis** [reply]
@manuschillerdev it is common to the body be a `Stream`, this means that when the request starts, not necessarily all data was sent to the server from the client. Also, the server can choose to read or not the request's content. Imagine if the body was too large, the server can check it for the `Content-Length` and refuse to read it.

**Reply by perenstrom** [reply]
> `req` is `http.IncomingMessage` so I have managed to read the body data by using a listener:
> 
> ```ts
> export const getServerSideProps: GetServerSideProps = async ({req}) => {
>   if (req.method == "POST") {
>     let body = ''
>     req.on('data', (chunk) => {
>       body += chunk
>     })
>     req.on('end', () => {
>       console.log(body);
>   });
>   }
>   return { props: {} };
> };
> ```

Hi! Thanks for your solution, works a charm. However, I have a bit of a hard time wrapping my head around getting the result from whatever I do with the request to my NextJS Page props. From what I understand, the on-end-listener is called whenever the request has finished incoming, and my getServerSideProps method is done before that?

Is there a way of awaiting the end-listener to be able to return the props I need, or should I go the route of `raw-body`?

**Reply by lvpaspaltzis** [reply]
go for the `raw-body`. It is easier and cleaner

**Reply by sametweb** [reply]
Here is another solution https://dev.to/smeijer/handling-post-requests-in-next-js-getserversideprops-50ia

**Reply by fadedlamp42** [reply]
@sametweb 's solution is

# SO GOOD

**Reply by omer-ayhan** [reply]
also here's another solution if you want to send a JSON response to the POST request instead of HTML 

```typescript
export async function getServerSideProps({ req, res }: GetServerSidePropsContext) {
 let body = '';
 if (req.method == 'POST') {
  req.on('data', (chunk) => {
   body += chunk;
  });
  res.statusCode = 200;
  res.setHeader('Content-Type', 'application/json');
  res.end(JSON.stringify({ message: 'successfully received notification' }));
  req.on('end', () => {
   console.log('\x1b[32mresponse\x1b[0m', body);
  });
 }

 return {
  props: {},
 };
}
```

**Comment by oentoro** [comment]
use formadible: 

[https://stackoverflow.com/questions/66674834/how-to-read-formdata-in-nextjs/67177153#67177153](https://stackoverflow.com/questions/66674834/how-to-read-formdata-in-nextjs/67177153#67177153)

```
import formidable from 'formidable'

export async function getServerSideProps( context ) {
  const { req } = context;
  const data = await new Promise((resolve, reject) => {
    const form = formidable()
    form.parse(req, (err, fields, files) => {
      if (err) reject({ err })
      resolve({ err, fields, files })
    }) 
  })
  console.log(data.fields)
  return {
    props: {},
  };
}

const Page = () => {
  return (
    <>
      <h1>Hello world</h1>
    </>
  );
};

export default Page;
```

**Comment by coheia** [comment]
I'm having issues with CORS, I'm trying to do a page visualization for the admin painel, the admin panel will send a request to the product page passing the data via body POST request and in the getServerSideProps I'll catch that and return the bodyData as a props instead of catch the product by API call...

But, in the Admin Panel, I'm getting CORS...
